padding scheme text (#209)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

See {{hpke-map}} for information on how a cipher suite maps to corresponding
HPKE algorithm identifiers.

minimum_inner_length
:  The minimum (encoded) length ClientHelloInner recommended by the server.
Sending a ClientHelloInner that is shorter than this will likely stand out.
Clients can use a padding extension within the ClientHelloInner to ensure

> the plaintext is at leat this long.

Reviewer: ```suggestion
the plaintext is at least this long.
```

See {{hpke-map}} for information on how a cipher suite maps to corresponding
HPKE algorithm identifiers.

maximum_name_length  
: The largest name the server expects to support, if known.
If this value is not known (e.g. if wildcard names are in use, or names can be

> added or removed from tne anonymity set during the lifetime of a particular

Reviewer 2: ```suggestion
added or removed from the anonymity set during the lifetime of a particular
```

 - an "echo_nonce" extension
 - TLS padding {{!RFC7685}}

An encoded ClientHelloInner can be around 300 octets long, but variations in
the length of the ciphertext version of that could defeat the entire purpose of
ECHO, if for example those expose the chosen server_name field in the
ClientHelloInner.  The padding ClientHello extension, if well used, can ensure
that length information does not expose what is contained in the
ClientHelloInner. Clients MUST add padding to the ClientHelloInner to try meet
this requirement.

Given that new extensions may be defined in future that contain sensitive

> length fields, we cannot describe all the ways in which length information

Reviewer: ```suggestion
Given that extensions could be defined in the future that reveal sensitive information
through their length, we cannot describe all the ways in which length information
```

could expose sensitive content. Implementers ought therefore be aware that they
might have to change their padding scheme as the set of supported extensions
changes. In addition to padding the ClientHelloInner, clients and servers MUST
both pad all other handshake messages that have sensitive length fields. For
example, if the client proposes ALPN values in the ClientHelloInner, the
server-selected value will be returned in an EncryptedExtension, so that
handshake message needs to be padded with TLS record layer padding. 

One approach that may be relative safe and simple is once ECHO is in use:

- When constructing the ClientHelloInner, use padding bytes so that the length

>   of the plaintext erncoding of ClienHelloInner is greater than

Reviewer: ```suggestion
  of the plaintext encoding of ClienHelloInner is greater than
```

ECHOConfig.minimum_inner_length and is a multiple of 16
- Both clients and servers pad all handshake messages so that plaintexts are a
  multiple of 16 octets long.
- Clients or servers MAY send additional padding bytes. This MUST NOT be

>   treated as an error.

Reviewer: What about the converse?  Is it an error to send something that is not padded to the minimum length?

Author: I think I'd say no, that's maybe better not treated as an error. But I could be convinced otherwise.

 - an "echo_nonce" extension
 - TLS padding {{!RFC7685}}

Variations in the length of the ciphertext version of the ClientHelloInner
could defeat the purpose of ECHO, if those expose the chosen server_name field
in the ClientHelloInner. The padding ClientHello extension, if well used, can
ensure that length information does not expose what is contained in the
ClientHelloInner. Clients SHOULD add padding to the ClientHelloInner to meet this
requirement.

Given that extensions could be defined in the future that reveal sensitive information
through their length, we cannot describe all the ways in which length information
could expose sensitive content. Implementers ought therefore be aware that they
might have to change their padding scheme as the set of supported extensions
changes. 

For each field in the inner ClientHello, clients need to determine how much to
pad given the semantics of that field.  For example, if a client can propose
various ALPN values, it could add padding to round up to the longest of those.

> For most fields in a ClientHello this can be determined without server help.

Reviewer 2: Can we also merge this paragraph with the preceding one?

> For the server_name, however, if ECHOConfig.maximum_name_length is longer than
> the actual server_name then clients SHOULD add padding to make up that
> difference. If the maximum_name_length is zero or less than the length of the
> actual server_name then round the server_name up to a multiple of 32 octets and
> randomly add another 32 octets 50% of the time and then include that amount
> of additional padding.

Reviewer 2: ```suggestion
However, the "server_name" extension could benefit from server input (ECHOConfig.maximum_name_length). 
Clients SHOULD compute the padding for this extension as follows:

1. If ECHOConfig.maximum_name_length is longer than the actual server_name then 
clients SHOULD add padding to make up that difference.
2. Otherwise, if ECHOConfig.maximum_name_length is zero or less than the length 
of the actual server_name then round the server_name up to a multiple of 32 octets 
and randomly add another 32 octets 50% of the time and then include that amount
of additional padding.
```

> Sum all the padding together, then in order to reduce entropy across different
> client implementationss, round up so that the inner ClientHello encoding length 
> is a multiple of 32.

Reviewer 2: ```suggestion
The amount of padding applied to the ClientHello is then computed as the sum of 
all per-extension padding values, rounded up to the nearest multiple of 32. 
(This additional rounding step aims to hide variance across different client
implementations.)
```

> In addition to padding the ClientHelloInner, clients and servers will also need
> to pad all other handshake messages that have sensitive-length fields. For
> example, if a client proposes ALPN values in the ClientHelloInner, the
> server-selected value will be returned in an EncryptedExtension, so that
> handshake message also needs to be padded using TLS record layer padding. 

Reviewer 2: ```suggestion
In addition to padding ClientHelloInner, clients and servers will also need
to pad all other handshake messages that have sensitive-length fields. For
example, if a client proposes ALPN values in ClientHelloInner, the
server-selected value will be returned in an EncryptedExtension, so that
handshake message also needs to be padded using TLS record layer padding.
```

entry whose CachedInformationType is "cert", since this indication
would divulge the true server name.

## Recommended Padding Scheme {#padding}

Future extensions could reveal sensitive information through their length. Consequently, 
padding should be flexible and support arbitrary extension changes. 

For each field in the inner ClientHello, clients need to determine how much to
pad given the semantics of that field.  For example, if a client can propose

> various ALPN values, it could add padding to round up to the longest of those.

Reviewer 3: Is this part right? Do clients need to consider padding for each field, or just the entire encrypted ClientHelloInner?

Reviewer 2: Yep -- this is what was discussed way up above in this PR.

Reviewer 3: Ok, I will re-read this PR, since I don't get it.

But, even so, I think this section needs a rationale. It says "clients need to determine how much to pad given the semantics of that field", but there is no reason given.

Author: I moved that rationale down to the security considerations section as suggested earlier. It could be moved back up, but basically anything where the length could be sensitive may need padding.

The target padding length of most ClientHello extensions can be determined without server help.
However, the "server_name" extension could benefit from server input (ECHOConfig.maximum_name_length). 
Clients SHOULD compute the padding for this extension as follows:

1. If ECHOConfig.maximum_name_length is longer than the actual server_name then 
clients SHOULD add padding to make up that difference.
2. Otherwise, if ECHOConfig.maximum_name_length is zero or less than the length 
of the actual server_name then round the server_name up to a multiple of 32 octets 
and randomly add another 32 octets 50% of the time and then include that amount

> of additional padding.

Reviewer 3: Suggest: "add at least another 32 octets".

Reviewer 4: Or "add one or more blocks of 32 octets" just to exclude behaviours that undo the effect of rounding the padding to a multiple of 32 octets.

Reviewer 2: 👍 to N >= 1 blocks. Clients can choose the distribution from which they want to sample N.

Reviewer 4: On second thoughts, we should probably match what's done in encrypted DNS, because otherwise we could create a potential leak from correlating the DNS request length with the ECHO length.

Author: Don't think copying DNS is correct. RFC 8467 calls for queries to be padded to a multiple of 128 octets and could be that the server-name is not the only sensitive name. 

I'm not sure there's a point in recommending against zero additional blocks of 32 - so long as you don't do zero too often it'd be ok. If you look back to the name lengths seen in the data I have it may be that zero additional blocks 50% of the time is ok. 

But mainly, I'd say we may be better to wait 'till we have a few implementations and then see what makes best sense.

Reviewer 2: > Don't think copying DNS is correct. RFC 8467 calls for queries to be padded to a multiple of 128 octets and could be that the server-name is not the only sensitive name.

Can you show that there is no information leakage if the two aren't in sync?

> I'm not sure there's a point in recommending against zero additional blocks of 32 - so long as you don't do zero too often it'd be ok. If you look back to the name lengths seen in the data I have it may be that zero additional blocks 50% of the time is ok.

The proposal above suggests N >= 1, not zero. In any case, as you suggest, spinning our wheels on something for which there is no backing data or analysis (yet) seems premature.

Reviewer 2: Hmm, is this resolved?

Reviewer 5: I actually don't think we should recommend the random padding. Let's do the minimum here.

The amount of padding applied to the ClientHello is then computed as the sum of 
all per-extension padding values, rounded up to the nearest multiple of 32. 
(This additional rounding step aims to hide variance across different client
implementations.)

In addition to padding ClientHelloInner, clients and servers will also need
to pad all other handshake messages that have sensitive-length fields. For
example, if a client proposes ALPN values in ClientHelloInner, the
server-selected value will be returned in an EncryptedExtension, so that

> handshake message also needs to be padded using TLS record layer padding.

Reviewer 3: Is there a good reference for "TLS record layer padding"? I think I know what this means, but it could be clearer.

Reviewer 2: It's defined in Section 5.4 of RFC8446: https://tools.ietf.org/html/rfc8446#section-5.4

Reviewer 3: Maybe that reference suggests a terminology issue. It says "All encrypted TLS records can be padded to inflate the size of the TLSCiphertext." Does ECHO define things in terms of TLSCiphertext?

I don't necessarily think there is a problem here, but in the two code bases I've read (rustls and NSS), the encrypted SNI or ClientHello haven't fit in all that cleanly.

odd, and probably some precise rules about handling ECHO and no
SNI uniformly?]]

## Padding

> Variations in the length of the ciphertext version of the ClientHelloInner

Reviewer 3: Suggest: "Variations in the length of the ClientHelloInner ciphertext..."

odd, and probably some precise rules about handling ECHO and no
SNI uniformly?]]

## Padding

Variations in the length of the ciphertext version of the ClientHelloInner

> could defeat the purpose of ECHO, if those expose the chosen server_name field

Reviewer 4: Do we need to update GREASE if we are rounding the packet sizes to multiples of 32 octets? If we want GREASE to look like ECHO then it should have the same external properties.

Reviewer 2: Yep -- there's a TODO in the GREASE section.
